frontend/src/metabase/query_builder/containers/QueryBuilder.unit.spec.tsx

```diff
@@ -502,12 +530,81 @@ describe("QueryBuilder", () => {
   });
 
   describe("unsaved changes warning", () => {
+    describe("creating models", () => {
+      afterEach(() => {
+        jest.clearAllMocks();
+      });
+
+      it("shows custom warning modal when leaving via SPA navigation", async () => {
+        const { history } = await setup({
+          card: null,
+          initialRoute: "/",
+        });
+
+        history.push("/model/new");
+        await waitForLoaderToBeRemoved();
+
+        await startNewNotebookModel();
+
+        history.goBack();
+
+        expect(screen.getByTestId("leave-confirmation")).toBeInTheDocument();
+      });
+
+      it("shows custom warning modal when leaving via Cancel button", async () => {
+        await setup({
+          card: null,
+          initialRoute: "/model/new",
+        });
+
+        await startNewNotebookModel();
+
+        userEvent.click(screen.getByRole("button", { name: "Cancel" }));
+
+        expect(screen.getByTestId("leave-confirmation")).toBeInTheDocument();
+      });
+
+      it("does not show custom warning modal when saving new model", async () => {
+        await setup({
+          card: null,
+          initialRoute: "/model/new",
+        });
+
+        fetchMock.get("path:/api/collection/1", [createMockCollection()]);
```

Comment:
I know @ranquild would ask you to use out setupMock methods

Reply:
It required a bit of refactoring, but here it is: 1645cdb702058cb799d0adbd7d144b8ed99ba1c7